### PR TITLE
fix: protect against phishing domain redirects in main/sub frames for http(s) requests

### PR DESCRIPTION
**Pull request by davidmurdoch**
[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/25153?quickstart=1)

### Easy: Verify non-redirect domains still work

1. Install the extension (you should test mv2 and mv3)
2. go through onboarding if necessary
2. go to a phishing website, like `claimblast.lol` (these websites seem to go down all the time, so you might have to grab one from https://github.com/MetaMask/eth-phishing-detect/commits/main/ to test with)
3. make sure it redirects to the phishing website

### Tricky: Verify domains that redirect are now blocked:

1. modify background.js by replacing `!phishingTestResponse?.result` with `hostname !== "shorturl.at"` (since we test blocking my test URL that is on a domain that isn't supposed to be blocked)
2. run `yarn start` (test both mv2 and mv3) 
3. install the extension
4. go through onboarding if necessary
4. navigate to `https://shorturl.at/6F2Zt`
5. you should see our phishing page. if you see evil.com it didn't work :-(

Some notes:

* If MetaMask has not fully initialized yet (like on a browser restart) redirect blocking won't work.
* The new phishing detection code kicks in _before_ our current `content_script`-based blocking, so it works for redirect cases as well as normal phishing blocking.

### CODE REVIEW

File: app/scripts/lib/local-store.js [hunk 1]
```diff
@@ -78,7 +78,7 @@ export default class ExtensionStore {
    * Returns all of the keys currently saved
    *
    * @private
-   * @returns {object} the key-value map from local storage
+   * @returns {Promise<object>} the key-value map from local storage
```
Comment (davidmurdoch) [inline app/scripts/lib/local-store.js:81]:
```suggestion
   * @returns {object} the key-value map from local storage
```

File: app/scripts/metamask-controller.js [hunk 2]
```diff
@@ -4784,8 +4784,27 @@ export default class MetamaskController extends EventEmitter {
    * @param {string} [options.subjectType] - The type of the sender, i.e. subject.
    */
   setupUntrustedCommunication({ connectionStream, sender, subjectType }) {
-    const { completedOnboarding } = this.onboardingController.store.getState();
-    const { usePhishDetect } = this.preferencesController.store.getState();
+    if (sender.url) {
+      if (this.onboardingController.store.getState().completedOnboarding) {
+        if (this.preferencesController.store.getState().usePhishDetect) {
```
Comment (davidmurdoch) [inline app/scripts/metamask-controller.js:4789]: This logic changed at one point, refactored, and finally restored to the original logic, but I kept some of the refactoring.

All the refactoring does is minimize the amount of "work" done in these pre-condition checks.

I don't mind restoring it to the original, as the refactoring's performance improvements here are certainly insignificant.

File: app/scripts/background.js [hunk 3]
```diff
@@ -176,6 +178,103 @@ const sendReadyMessageToTabs = async () => {
...
+      }
+
+      // ignore requests that come from our phishing warning page
+      if (
+        details.initiator &&
+        // compare normalized URLs
+        new URL(details.initiator).toString() === phishingPageHref
+      ) {
+        return {};
+      }
```
Comment (davidmurdoch) [inline app/scripts/background.js:226]: This section just short-circuits if the preconditions aren't met. Most of the conditions are the same as the checks in `setupUntrustedCommunication`, except for the addition of `details.initiator`.

File: app/scripts/background.js [hunk 4]
```diff
@@ -176,6 +178,103 @@ const sendReadyMessageToTabs = async () => {
...
+
+      const { hostname, href, searchParams } = new URL(details.url);
+      if (inTest) {
+        if (searchParams.has('IN_TEST_BYPASS_EARLY_PHISHING_DETECTION')) {
+          // this is a test page that needs to bypass early phishing detection
+          return {};
+        }
+      }
```
Comment (davidmurdoch) [inline app/scripts/background.js:234]: I'm not proud of the `IN_TEST` logic sprinkled in here, but I'm not sure if there is a better way. Let me know if you think of one!

File: app/scripts/background.js [hunk 5]
```diff
@@ -176,6 +178,103 @@ const sendReadyMessageToTabs = async () => {
...
+          redirectTab(details.tabId, redirectHref);
+          // don't let the sub_frame load at all
+          return { cancel: true };
+        }
+        // redirect the whole tab, even if it's a sub_frame request
+        return { redirectUrl: redirectHref };
+      }
+      // redirect the whole tab, even if it's a sub_frame request
+      redirectTab(details.tabId, redirectHref);
+      return {};
```
Comment (davidmurdoch) [inline app/scripts/background.js:271]: If main frame:

* mv2: immediately redirects to our Phishing Warning Page. no requests are made to the phishing server
* mv3: a redirect to our phishing warning page is initiated and completed asynchronously. Any redirects from the phishing page may be followed, but ultimately end up being overridden by the redirect to our Phishing Warning Page.

If sub frame (iframe):

* mv2: cancels the iframe's navigation. No requests are made to the phishing server. A redirect to our Phishing Warning Page is made asynchronously in the main frame (the tab).
* mv3: a redirect to our phishing warning page is initiated in the main frame (the tab) and completed asynchronously. Any redirects from the phishing page may still be followed, but ultimately end up being overridden by the redirect to our Phishing Warning Page.

File: test/e2e/tests/phishing-controller/phishing-detection.spec.js [hunk 6]
```diff
@@ -297,4 +330,151 @@ describe('Phishing Detection', function () {
...
+            <meta http-equiv="Refresh" content="0;url="${destination}"/>
+            <title>Phishing test</title>
+
+            <script>
+              // this script should not run.
+              // it is meant to test for regressions in our redirect
+              // protection due to changes in either MetaMask or browsers.
+              document.location.href = "${destination}";
+              alert("trying to prevent phishing protection");
+              while(true){}
```
Comment (davidmurdoch) [inline test/e2e/tests/phishing-controller/phishing-detection.spec.js:391]: Random ideas to try to work around our phishing protections. They don't affect our protections and are actually not necessary for our testing. I figured I could keep them here for regression testing. Thoughts?

File: app/manifest/v2/_base.json [hunk 7]
```diff
@@ -64,14 +64,11 @@
...
-    "https://*.infura.io/",
-    "https://*.codefi.network/",
-    "https://*.cx.metamask.io/",
-    "https://chainid.network/chains.json",
-    "https://lattice.gridplus.io/*",
+    "http://*/*",
+    "https://*/*",
     "activeTab",
     "webRequest",
+    "webRequestBlocking",
```
Comment (davidmurdoch) [inline app/manifest/v2/_base.json:71]: The question: does this cause permissions warnings to FF users (and any other MV2 browsers we build for)?

File: development/build/set-environment-variables.js [hunk 8]
```diff
@@ -189,28 +189,24 @@ function getPhishingWarningPageUrl({ variables, testing }) {
         }/`;
   }
 
-  // We add a hash/fragment to the URL dynamically, so we need to ensure it
-  // has a valid pathname to append a hash to.
-  const normalizedUrl = phishingWarningPageUrl.endsWith('/')
-    ? phishingWarningPageUrl
-    : `${phishingWarningPageUrl}/`;
```
Comment (davidmurdoch) [inline development/build/set-environment-variables.js:196]: This was error prone if the phishing page ends up ever being something like `https://domain.com/page-with-trailing-slash.html`. The logic is replaced with `phishingWarningPageUrlObject.toString();` below, which adds a trailing slash to the end of the domain if it is missing and there URL has nothing after it.

File: test/e2e/tests/phishing-controller/phishing-detection.spec.js [hunk 9]
```diff
@@ -297,4 +330,151 @@ describe('Phishing Detection', function () {
...
+          resolve(fixtures.driver); // resolve this `beforeEach`
+          await deferredTestSuite.promise; // now wait for all tests to complete
+        },
+      );
+      driver = await promise;
+
+      // required to ensure MetaMask is fully started before running tests
+      // if we had a way of detecting when the offscreen/background were ready
+      // we could remove this
+      await unlockWallet(driver);
```
Comment (davidmurdoch) [inline test/e2e/tests/phishing-controller/phishing-detection.spec.js:447]: sad day.

I couldn't easily set up tests to wait until the background/service_worker was fully ready, so I had to fall back to the `unlockWallet` strategy.

Note: the wallet does NOT have to be unlocked for the protections to kick in.

File: test/e2e/tests/phishing-controller/phishing-detection.spec.js [hunk 10]
```diff
@@ -297,4 +330,151 @@ describe('Phishing Detection', function () {
       },
     );
   });
+
+  describe('Phishing redirect protections', function () {
```
Comment (davidmurdoch) [inline test/e2e/tests/phishing-controller/phishing-detection.spec.js:334]: This test is not structured like our other tests simple for "best practice" reasons: tests set up _should_ be done in `before`/`beforeAll`, and teardown should be done in `after`/`afterAll`.

By making use of `before`/`beforeAll` this test sets up the browser fixtures only once for all of the tests. It still uses out `withFixtures` function, so it gets a bit awkward because this helper function requires an async callback and it returns `Promise` (for different and valid reasons, but you already know that).

Anyway, I don't mind redoing it if you hate it this way. I don't even particularly love the code-gymnastics I had to do to make `withFixtures` work in this test flow. It does make it faster though.

File: app/scripts/background.js (hunk 4 above)
Comment (davidmurdoch) [reply app/scripts/background.js:234]: It can get into this state naturally. The `browser.webRequest.onBeforeRequest` hook takes a bit before the browser "installs" it on start up. But our `content_script` looks to be injected immediately. This flag is meant to test that case. Under "natural" use this is just a race condition, but we can't rely on that for our tests, so this code block is mean to simulate it, and forces the `content_script` to do the phishing detection.

File: test/e2e/tests/phishing-controller/phishing-detection.spec.js (hunk 9 above)
Comment (davidmurdoch) [reply test/e2e/tests/phishing-controller/phishing-detection.spec.js:447]: I've figured out a way in a separate PR meant to fix a flaky test.

@HowardBraham figured out a way to set up a socket connection from the extension to the test runner in https://github.com/MetaMask/metamask-extension/pull/25362, so we might be able to use that (which I think will be even better than my other solution)!

File: app/scripts/background.js [hunk 11]
```diff
@@ -176,6 +178,103 @@ const sendReadyMessageToTabs = async () => {
...
+      const prefState = theController.preferencesController.store.getState();
+      if (!prefState.usePhishDetect) {
+        return {};
+      }
+
+      // ignore requests that come from our phishing warning page
+      if (
+        details.initiator &&
+        // compare normalized URLs
+        new URL(details.initiator).toString() === phishingPageHref
```
[the remark this answers is not in the document]
Comment (davidmurdoch) [reply app/scripts/background.js:223]: Oh interesting. I'll test it.

IIRC, this check was necessary because the "Continue to site" link on our phishing warning page begins the redirect before our `content_script` can tell the `phishingController` that the domain is safe.

[the remark this answers is not in the document]
Comment (davidmurdoch) [reply app/scripts/background.js:233]: I've tested this and while the `sandbox` property on the iframe did cause everything to stop working, it wasn't for the reason you'd expect.

So `sandbox` actually prevents our `Continue to site` links (and other links that want to open in `_blank` like "Open this warning in a new tab") from working at all.

This is an existing issue that I think is out of scope for this PR. I _think_ the solution is for the iframe to send a message to background.js asking it to open the new tab for us (since the `sandbox` attribute without an `allow-popups` prevents it).

I encourage reviewers to attempt their own manual testing with `<iframe sandbox>`.

**Comment by seaona** [comment]
## QA findings (in progress)
So far looks good. I've encountered one minor change which might not be expected: whenever I click to the `continue to the site` link from the MM Phishing page, I see how I remain in the MM page. Only after I click again for a second time, or when I refresh, I am then directed to the phishing page.
I've seen that this behaviour doesn't seem to happen in production, so I was wondering if might be due to some of the changes? :thinking: 

https://github.com/MetaMask/metamask-extension/assets/54408225/26ea726c-e637-4a65-b2b1-04ec2d047d65

**Comment by davidmurdoch** [comment]
> ## QA findings (in progress)
> So far looks good. I've encountered one minor change which might not be expected: whenever I click to the `continue to the site` link from the MM Phishing page, I see how I remain in the MM page. Only after I click again for a second time, or when I refresh, I am then directed to the phishing page. I've seen that this behaviour doesn't seem to happen in production, so I was wondering if might be due to some of the changes? 🤔
> 
>  double-click-to-go-to-page.mp4

Hm, I've seen this issue before and thought I had it solved. I'll investigate a fix for this tomorrow. Thanks!

**Comment by davidmurdoch** [comment]
> ## QA findings (in progress)
> So far looks good. I've encountered one minor change which might not be expected: whenever I click to the `continue to the site` link from the MM Phishing page, I see how I remain in the MM page. Only after I click again for a second time, or when I refresh, I am then directed to the phishing page. I've seen that this behaviour doesn't seem to happen in production, so I was wondering if might be due to some of the changes? 🤔
> 
>  double-click-to-go-to-page.mp4

@seaona I've committed a fix and some proof of testing: 

https://github.com/MetaMask/metamask-extension/assets/187813/84a74aaa-6eef-4bde-b724-c10b290869f1

File: app/manifest/v2/_base.json (hunk 7 above)
Comment (Gudahtt) [reply app/manifest/v2/_base.json:71]: Instructions here: https://extensionworkshop.com/documentation/develop/test-permission-requests/

I can test this

Comment (Gudahtt) [reply app/manifest/v2/_base.json:71]: I was able to test this. I created three branches:
* `manifest-update-test-1`
* `manifest-update-test-2`
* `manifest-update-test3`.

I pushed the manifest changes I made to those branches. I used a local update server, which I ran using the command `./development/static-server.js -p 9999 ./firefox-updates`. The `firefox-updates` directory was a local directory I created with the XPI files, and an `updates.json` file that had the relevant updates.

I used Firefox Developer Edition, but I had to follow some Nightly-specific instructions in that doc (I guess it's inaccurate in suggesting they're only needed on nightly). It was the instructions about setting things in `about:config` to enable a local update server.

Here was the final `updates.json` file I used:

```
{
  "addons": {
    "webextension@metamask.io": {
      "updates": [
        {
          "version": "11.16.7",
          "update_link": "http://localhost:9999/metamask-firefox-11.16.7.xpi"
        },
        {
          "version": "11.16.8",
          "update_link": "http://localhost:9999/metamask-firefox-11.16.8.xpi"
        },
        {
          "version": "11.16.9",
          "update_link": "http://localhost:9999/metamask-firefox-11.16.9.xpi"
        }
      ]
    }
  }
}
```

I created a build from each branch using `ENABLE_MV3 yarn list`. The first branch was based on `develop`, the second was based on this branch, and the third was also based on this branch but had the additional `tabs` permission (which should result in a warning on update). Each branch had the `update_url` manifest change and a version bump as well.

I first removed `11.16.9` from `updates.json` to test this branch, and no warning appeared. Then I added `11.16.9` back, and I did get a warning about the tabs permission being added.

We should be safe to add this permission, there was no additional warning.

File: app/scripts/background.js [hunk 12]
```diff
@@ -114,6 +114,8 @@ if (inTest || process.env.METAMASK_DEBUG) {
 }
 
 const phishingPageUrl = new URL(process.env.PHISHING_WARNING_PAGE_URL);
+// normalized the URL once and reuse it:
+const phishingPageHref = phishingPageUrl.toString();
```
Comment (Gudahtt) [inline app/scripts/background.js:118]: Nit: I see that this href is used to create a new URL instance twice later in this file. Could we use the original `phishingPageUrl` instead? It seems equivalent to me, unless we expect `phishingPageUrl` to be mutated (which would be surprising), or unless there is some normalization happening here that makes `phishingPageUrl` and `new URL(phishingPageHref)` different.

File: app/scripts/background.js [hunk 13]
```diff
@@ -184,6 +186,111 @@ const sendReadyMessageToTabs = async () => {
...
+      }
+
+      // ignore requests that come from our phishing warning page, as
+      // the requests may come from the "continue to site" link, so we'll
+      // actually _want_ to bypass the phishing detection. We shouldn't have to
+      // do this, because the phishing site does tell the extension that the
+      // domain it blocked it now "safe", but it does this _after_ the request
+      // begins (which would get blocked by this listener). So we have to bail
+      // on detection here.
+      // This check can be removed once  https://github.com/MetaMask/phishing-warning/issues/160
```
Comment (Gudahtt) [inline app/scripts/background.js:234]: Neat, good catch

File: app/scripts/background.js [hunk 14]
```diff
@@ -184,6 +186,111 @@ const sendReadyMessageToTabs = async () => {
...
+        details.initiator &&
+        // compare normalized URLs
+        new URL(details.initiator).host === phishingPageUrl.host
+      ) {
+        return {};
+      }
+
+      const { hostname, href, searchParams } = new URL(details.url);
+      if (inTest) {
+        if (searchParams.has('IN_TEST_BYPASS_EARLY_PHISHING_DETECTION')) {
```
Comment (Gudahtt) [inline app/scripts/background.js:246]: Interesting solution!

I was thinking that we could solve this by serving this inner page from a different origin. But then we'd need to update our phishing test code to take origins into consideration (or maybe just consider them for localhost? Or only if a port is present in the blocked site entry?). It just considers hostname for now.

This seems good enough though.

File: app/scripts/background.js [hunk 15]
```diff
@@ -184,6 +186,111 @@ const sendReadyMessageToTabs = async () => {
...
+
+      theController.phishingController.maybeUpdateState();
+      const phishingTestResponse =
+        theController.phishingController.test(hostname);
+      if (!phishingTestResponse?.result) {
+        return {};
+      }
+
+      theController.metaMetricsController.trackEvent({
+        // should we differentiate between background redirection and content script redirection?
```
Comment (Gudahtt) [inline app/scripts/background.js:260]: Hmm. I don't think it matters? I can't think of a reason to care about where we catch it.

File: app/scripts/background.js [hunk 16]
```diff
@@ -184,6 +186,111 @@ const sendReadyMessageToTabs = async () => {
...
+      // the browser from loading the page at all
+      if (isManifestV2) {
+        if (details.type === 'sub_frame') {
+          // redirect the entire tab to the
+          // phishing warning page instead.
+          redirectTab(details.tabId, redirectHref);
+          // don't let the sub_frame load at all
+          return { cancel: true };
+        }
+        // redirect the whole tab, even if it's a sub_frame request
```
Comment (Gudahtt) [inline app/scripts/background.js:282]: Nit: This comment seems a bit misleading. It says "even if it's a sub_frame request", but sub_frame requests are handled in the block above right? So we would never get here.

Though I guess that block also redirects the whole tab, so it's technically true. But I read this to imply the line below is the part doing that, but it's not.

File: app/scripts/background.js (hunk 12 above)
Comment (Gudahtt) [reply app/scripts/background.js:118]: Oh I see this comment about this topic: https://github.com/MetaMask/metamask-extension/pull/25153/files#r1639008401

Perhaps helpful to update this comment to explain that this will "add a trailing slash to the end of the domain if it is missing".

Comment (Gudahtt) [reply app/scripts/background.js:118]: I guess it shouldn't be missing, because of that normalization happening in the build script there 😅 But I guess it couldn't hurt to be doubly sure.

File: test/e2e/tests/phishing-controller/phishing-detection.spec.js [hunk 17]
```diff
@@ -1,4 +1,7 @@
+/* eslint-disable mocha/no-hooks-for-single-case, mocha/no-sibling-hooks, no-loop-func */
```
Comment (Gudahtt) [inline test/e2e/tests/phishing-controller/phishing-detection.spec.js:1]: Could you add a brief explanation here for why these rules have been disabled? The reasons you stated below do make sense, but we can say that here in a comment. I don't want to normalize the behavior of disabling ESLint rules without explanation because people tend to do it for the wrong reasons, without understanding what the rule is for.

It might be easier to move the `no-loop-func` to the line it applies to as well, since there's only one violation. The explanation would make more sense alongside the relevant code.

Comment (davidmurdoch) [reply test/e2e/tests/phishing-controller/phishing-detection.spec.js:1]: I slapped them up here because eslint rule comments are so noisy and make code harder to read (I also think many of our eslint rules themselves neccessitate writing code that is harder to read in the first place, but I digress).

I won't die on this hill, so I'll make the changes as requested.

Comment (Gudahtt) [reply test/e2e/tests/phishing-controller/phishing-detection.spec.js:1]: `no-loop-func` is a frustrating one. Generally it's a good rule in that it can catch some nasty errors, but it can't tell the difference between the problematic cases (where a reader might wrongly expect a variable to have the same value inside the function as it does when the function is declared in the loop), and the innocuous cases like this (references that aren't specific to a given iteration).

I would say something like
```
      // This rule is flagging unsafe references to `server` and `driver`, but they are being used
      // safely here. We are intentionally sharing one instance for each test.
      // eslint-disable-next-line no-loop-func
```

For the other two, it maybe suffices to say that you've broken those rules to improve readability.

File: app/scripts/background.js (hunk 14 above)
Comment (Gudahtt) [reply app/scripts/background.js:246]: Oh sorry I misunderstood, I thought this was a workaround to enable testing a "redirect from a blocked site to an unblocked site" using purely local sites. This is just to test one solution over the other in a reliable way. Nevermind then!

File: test/e2e/tests/phishing-controller/phishing-detection.spec.js [hunk 18]
```diff
@@ -297,4 +332,151 @@ describe('Phishing Detection', function () {
...
+      if (server) {
+        const { promise, resolve } = createDeferredPromise();
+        server.close(() => resolve());
+        // We need to close all connections to stop the server quickly
+        // Otherwise it takes a few seconds for it to close
+        server.closeAllConnections();
+        await promise;
+      }
+    });
+    const deferredTestSuite = createDeferredPromise();
```
Comment (Gudahtt) [inline test/e2e/tests/phishing-controller/phishing-detection.spec.js:421]: Haha, wow that is an interesting solution to that problem.

It does make sense to enable using a single `withFixtures` call for multiple tests. Have you tried moving the tests inside the `withFixtures` callback instead? That might be a simpler way to do this.

**Review by Gudahtt — COMMENTED** [review]
Looks great! I have nearly finished review but I need to finish reading the tests, I didn't get through all of that. I will return to this next week.

File: test/e2e/tests/phishing-controller/phishing-detection.spec.js (hunk 17 above)
Comment (Gudahtt) [reply test/e2e/tests/phishing-controller/phishing-detection.spec.js:1]: >I also think many of our eslint rules themselves neccessitate writing code that is harder to read in the first place, but I digress

Please follow up in Slack with specifics! We should have rules that encouraging writing easier to read code. Though to some degree that will be subjective. Maybe some of the things you are annoyed with can be reconsidered.

>eslint rule comments are so noisy and make code harder to read

To clarify, I don't generally object to file-wide ignore comments, I do think that would be better for the ones regarding test hooks in this case.

For the `no-loop-func` one, it seemed easier to understand that one in-context but whatever you would prefer is fine by me, as long as there's at least a one-line comment somewhere on why we're disabling it.

File: test/e2e/tests/phishing-controller/phishing-detection.spec.js (hunk 18 above)
Comment (davidmurdoch) [reply test/e2e/tests/phishing-controller/phishing-detection.spec.js:421]: Yes, but then we'd be back to doing things The Wrong Way™️.

Test set up should be inside `before` or `beforeAll` and tear down in `after` or `afterAll`. Doing it the Right Way®️ will give us access to more granular test metrics and proper setup/tear down will prevent cascading test failures due to issue with open handles (when the test run kills a the test process for whatever reason, it _always_ runs the after hooks).

Perhaps `withFixtures` should be refactored to be used in it's current form, as well as in a form that makes the patterns in this PR less chaotic; maybe a `startFixtures` function that returns something like `Promise<{ driver, <... etc...>, end: () => Promise<void> }>`  where `end` could be called by the `after`/`afterAll`. Thoughts (and if yes, should it be done in this PR, before this PR, or as a follow up PR)?

File: app/scripts/background.js (hunk 14 above)
Comment (davidmurdoch) [reply app/scripts/background.js:246]: I actually really hate putting testing-only logic in the runtime code. If you can think of a better way to do this do let me know!

File: app/scripts/background.js (hunk 15 above)
Comment (davidmurdoch) [reply app/scripts/background.js:260]: Yeah, I wasn't sure. Maybe someone would want to know how often we're redirecting from the content script vs this earlier detection?

File: app/scripts/background.js (hunk 16 above)
Comment (davidmurdoch) [reply app/scripts/background.js:282]: Yup, an outdated comment. I'll update it

File: app/scripts/background.js (hunk 12 above)
Comment (davidmurdoch) [reply app/scripts/background.js:119]: I've added a clarifying comment.

File: app/scripts/background.js (hunk 16 above)
Comment (davidmurdoch) [reply app/scripts/background.js]: Done!

File: test/e2e/tests/phishing-controller/phishing-detection.spec.js (hunk 17 above)
Comment (davidmurdoch) [reply test/e2e/tests/phishing-controller/phishing-detection.spec.js]: Done!

File: app/manifest/v2/_base.json [hunk 19]
```diff
@@ -64,14 +64,11 @@
...
     "unlimitedStorage",
     "clipboardWrite",
-    "http://localhost:8545/",
-    "https://*.infura.io/",
-    "https://*.codefi.network/",
-    "https://*.cx.metamask.io/",
-    "https://chainid.network/chains.json",
-    "https://lattice.gridplus.io/*",
+    "http://*/*",
+    "https://*/*",
```
[the remark this answers is not in the document]
Comment (davidmurdoch) [reply app/manifest/v2/_base.json:68]: I don't believe so.

[the remark this answers is not in the document]
Comment (davidmurdoch) [reply app/manifest/v2/_base.json:68]: That said, we already practically have these `allow-all` permissions via the `content_scripts` `matches`. However, content_script only works _after_ redirects have settled, so they aren't useful.

Not sure it makes it any better, but the MV3 manifest does already include the same `*` permissions:

```
  "host_permissions": [
    "http://localhost:8545/",
    "file://*/*",
    "http://*/*",
    "https://*/*"
  ],
  ```

File: test/e2e/tests/phishing-controller/phishing-detection.spec.js (hunk 18 above)
Comment (Gudahtt) [reply test/e2e/tests/phishing-controller/phishing-detection.spec.js]: Hmm, our general preference in other test suites has been to use "setup functions" over hooks for setup steps, so that they can be configured on a per-test basis while still reusing the setup code. For example, that's what `withFixtures` does.  Another example of this is the `withController` setup function we have in various controller unit test suites. Our unit test contributor docs suggest this pattern as well (see [here](https://github.com/MetaMask/contributor-docs/blob/main/docs/unit-testing.md#avoid-the-use-of-beforeeach), in the "Read more" section)

This would still include "proper setup/tear down" because we can tear down in a `finally` block in the setup function.

Placing multiple tests within a `withFixtures` callback does break test isolation, but so does what you're doing here. And that seems like a reasonable tradeoff to make in an e2e test context.

tl;dr, maybe putting multiple tests in a `withFixtures` callback wouldn't be "The Wrong Way™️"?

I don't know what "granular test metrics" means in this context though, maybe there is some benefit to using hooks that I am missing there.

File: test/e2e/tests/phishing-controller/phishing-detection.spec.js [hunk 20]
```diff
@@ -60,45 +65,74 @@ describe('Phishing Detection', function () {
...
+
+    it('should display the MetaMask Phishing Detection page in an iframe and take the user to the blocked page if they continue', async function () {
+      await withFixtures(
+        getFixtureOptions({
+          title: this.test.fullTitle(),
+          dappPaths: [
+            './tests/phishing-controller/mock-page-with-iframe-but-disable-early-detection',
+          ],
+        }),
+        getTest(true),
```
Comment (Gudahtt) [inline test/e2e/tests/phishing-controller/phishing-detection.spec.js:133]: Nit: Maybe better to use named parameters here, as `getText({ expectIframe: true })` is easier to read than `getText(true)`

**Review by Gudahtt — APPROVED** [review]
LGTM, great work!

Lots of room to improve the readability of the test suite, but we can do that later

File: test/e2e/tests/phishing-controller/phishing-detection.spec.js (hunk 18 above)
Comment (davidmurdoch) [reply test/e2e/tests/phishing-controller/phishing-detection.spec.js]: > we can tear down in a finally block in the setup function.

There are cases where a test can be forcibly killed by the test runner (mocha, jest) and `finally` blocks are _not_ run, but `after` will be.

Not doing test cleanup is not desirable (I know you know, but others might read this too, so I'm saying it anyway), as it can leave files behind or port-bound processes running, like ganache or other servers -- and the next test that attempts to spin up these servers will also fail. This makes understanding debug logs a bit more cumbersome.

I've thought of two cases I've seen happen in the extension code base.

When a test times out a `finally` won't run:

```
// name this file test.js then run `npx -y mocha test.js`
describe("test", () => {
  before(()=> {
    console.log("before");
  });

  after(()=> {
    console.log("after");
  });

  function runTest() {
    return new Promise(resolve => {
      setTimeout(resolve, 10000);
    })
  }

  it("works", async () => {
    try {
      await runTest()
    } finally {
      console.log("finally"); // this never runs
    }
  })
})
```

Another case where `finally` is not sufficient is when an unhandled exception/rejection bubbles up to the top process. Example:

```javascript
// name this file test.js then run `npx -y mocha test.js`
describe("test", () => {
  before(()=> {
    console.log("before");
  });

  after(()=> {
    console.log("after");
  });

  function runTest() {

    setTimeout(() => {
      // pretend this is thrown by some listener we've got installed somewhere
      throw new Error("error");
    }, 100);


    return new Promise(resolve => {
      setTimeout(resolve, 1000);
    })
  }

  it("works", async () => {
    try {
      await runTest()
    } finally {
      console.log("finally"); // this will not be logged
    }
  })
})
```

Regarding granular timings: we can measure how long the actual tests take, vs the test setup and teardown. We can then more easily optimize for them. Not like it's useful right now, as we've got way bigger fish to fry, but if we wanted to start implementing upper bounds for test duration these timings could prove useful.

**Comment by davidmurdoch** [comment]
> LGTM, great work!
> 
> Lots of room to improve the readability of the test suite, but we can do that later

I've made a new issue: https://github.com/MetaMask/metamask-extension/issues/25667 and assigned myself to it. I'll be OOO for a while, but I'll try to get it cleaned up when I return!

File: test/e2e/tests/phishing-controller/phishing-detection.spec.js (hunk 20 above)
Comment (davidmurdoch) [reply test/e2e/tests/phishing-controller/phishing-detection.spec.js:133]: Ah, good nit! I've bundled this suggestion in with https://github.com/MetaMask/metamask-extension/issues/25667#issue-2389530827

File: app/manifest/v2/_base.json (hunk 19 above)
[the remark this answers is not in the document]
Comment (Gudahtt) [reply app/manifest/v2/_base.json:68]: Due to the existing contentscript permissions, these additions don't really grant us much in terms of added capabilities. The only differences I am aware of are that it lets us use the `webRequest` API, and it disables CORS checks.

I would like to avoid adding further permissions, but, in this case it does seem like we have a good reason.

File: app/scripts/metamask-controller.js (hunk 2 above)
Comment (danjm) [reply app/scripts/metamask-controller.js:4772]: this refactor looks fine

File: development/build/set-environment-variables.js (hunk 8 above)
Comment (danjm) [reply development/build/set-environment-variables.js:196]: :+1:

**Review by danjm — APPROVED** [review]
LGTM
